src/Build/Utilities/NuGetFrameworkWrapper.cs

```diff
@@ -133,5 +156,106 @@ public string FilterTargetFrameworks(string incoming, string filter)
                 });
             }
         }
+
+#if FEATURE_APPDOMAIN
+        /// <summary>
+        /// A null-returning InitializeLifetimeService to give the proxy an infinite lease time.
+        /// </summary>
+        public override object InitializeLifetimeService() => null;
+
+        private static AppDomainSetup CreateAppDomainSetup(string assemblyDirectory)
+        {
+            string assemblyPath = Path.Combine(assemblyDirectory, NuGetFrameworksFileName);
+            AssemblyName assemblyName;
+            try
+            {
+                assemblyName = AssemblyName.GetAssemblyName(assemblyPath);
+            }
+            catch
+            {
+                // Return null to fall back to loading into the default AppDomain using LoadFile.
+                return null;
+            }
+
+            byte[] publicKeyToken = assemblyName.GetPublicKeyToken();
+            StringBuilder publicKeyTokenString = new(publicKeyToken.Length * 2);
+            for (int i = 0; i < publicKeyToken.Length; i++)
+            {
+                publicKeyTokenString.Append(publicKeyToken[i].ToString("x2", CultureInfo.InvariantCulture));
+            }
+
+            // Create an app.config for the AppDomain. We expect the AD to host the currently executing assembly Microsoft.Build,
```

Comment:
Is it possible to avoid the `Microsoft.Build` redirects by using the app domain only _within_ the `NuGetFrameworkWrapper`, instead of creating that inside the domain too?

Reply:
We need _an_ MBRO in the app domain to call through. It may as well live in `NuGet.Frameworks`, that's true. It would mean maintaining an internal API made specifically for MSBuild in another repo, and I'm not sure if it's worth the hassle. This part of the code has been relatively stable .. but so have been the binding redirects. I would probably err on the side of simplicity and keep the duplicated redirects here (could be extracted from our app.config using a source generator!). What do you think?

Reply:
Let's go with this and see if there's a reason to change it later.